main/src/ca/uwaterloo/flix/language/phase/Verifier.scala

```diff
@@ -235,6 +242,68 @@ object Verifier {
           check(expected = MonoType.Enum(sym.enumSym))(actual = t1, loc)
           tpe
 
+        case AtomicOp.ArrayLength =>
+          val List(arrt) = ts
+          arrt match {
+            case MonoType.Array(_) => check(expected = MonoType.Int32)(actual = tpe, loc)
+            case _ => throw MismatchedShape(tpe, s"An Array", loc)
+          }
+
+        case AtomicOp.ArrayNew =>
+          val List(elmt, lent) = ts
+          check(expected = MonoType.Int32)(actual = lent, loc)
+          check(expected = MonoType.Array(elmt))(actual = tpe, loc)
+
+        case AtomicOp.ArrayLit =>
+          tpe match {
+            case MonoType.Array(elmt) =>
+              ts.foreach(t => check(expected = elmt)(actual = t, loc))
+            case _ => throw MismatchedShape(tpe, s"An Array", loc)
+          }
+          tpe
+
+        case AtomicOp.ArrayLoad =>
+          val List(arrt, idxt) = ts
+          check(expected = MonoType.Int32)(actual = idxt, loc)
+          arrt match {
+            case MonoType.Array(elmt) => check(expected = elmt)(actual = tpe, loc)
+            case _ => throw MismatchedShape(tpe, s"An Array", loc)
+          }
+          tpe
+
+        case AtomicOp.ArrayStore =>
+          val List(arrt, idxt, expt) = ts
+          check(expected = MonoType.Int32)(actual = idxt, loc)
+          arrt match {
+            case MonoType.Array(elmt) => check(expected = elmt)(actual = expt, loc)
```

Comment:
you use `check` everywhere, but for nice errors its better to use `checkEq` in cases like this where both types are unknown - we don't know what the correct type is, just that they have to agree